test/powershell/Modules/Microsoft.PowerShell.Core/ModuleConstraint.Tests.ps1

```diff
@@ -0,0 +1,1838 @@
+
+$successCases = @(
+    @{
+        ModuleVersion = '2.0'
+        MaximumVersion = $null
+        RequiredVersion = $null
+    },
+    @{
+        ModuleVersion = '1.0'
+        MaximumVersion = '3.0'
+        RequiredVersion = $null
+    },
+    @{
+        ModuleVersion = $null
+        MaximumVersion = '3.0'
+        RequiredVersion = $null
+    },
+    @{
+        ModuleVersion = $null
+        MaximumVersion = $null
+        RequiredVersion = $actualVersion
+    }
+)
+
+$failCases = @(
+    @{
+        ModuleVersion = '2.5'
+        MaximumVersion = $null
+        RequiredVersion = $null
+    },
+    @{
+        ModuleVersion = '2.0'
+        MaximumVersion = '2.2'
+        RequiredVersion = $null
+    },
+    @{
+        ModuleVersion = '3.0'
+        MaximumVersion = '3.1'
+        RequiredVersion = $null
+    },
+    @{
+        ModuleVersion = '3.0'
+        MaximumVersion = '2.0'
+        RequiredVersion = $null
+    },
+    @{
+        ModuleVersion = $null
+        MaximumVersion = '1.7'
+        RequiredVersion = $null
+    },
+    @{
+        ModuleVersion = $null
+        MaximumVersion = $null
+        RequiredVersion = '2.2'
+    }
+)
+
+$guidSuccessCases = [System.Collections.ArrayList]::new()
+foreach ($case in $successCases)
+{
+    [void]$guidSuccessCases.Add($case + @{ Guid = $null })
+    [void]$guidSuccessCases.Add(($case + @{ Guid = $actualGuid }))
+}
+
+$guidFailCases = [System.Collections.ArrayList]::new()
+foreach ($case in $failCases)
+{
+    [void]$guidFailCases.Add($case + @{ Guid = $null })
+    [void]$guidFailCases.Add(($case + @{ Guid = $actualGuid }))
+    [void]$guidFailCases.Add(($case + @{ Guid = [guid]::NewGuid() }))
+}
+
+Describe "Module loading with version constraints" -Tags "Feature" {
+    BeforeAll {
+        $moduleName = 'TestModule'
+        $modulePath = Join-Path $TestDrive $moduleName
+        New-Item -Path $modulePath -ItemType Directory
+        $manifestPath = Join-Path $modulePath "$moduleName.psd1"
+        New-ModuleManifest -Path $manifestPath -ModuleVersion $actualVersion -Guid $actualGuid
+
+        $oldPSModulePath = $env:PSModulePath
+        $env:PSModulePath += [System.IO.Path]::PathSeparator + $TestDrive
+    }
+
+    AfterAll {
+        $env:PSModulePath = $oldPSModulePath
+    }
+
+    AfterEach {
+        Get-Module $moduleName | Remove-Module
+    }
+
+    It "Loads the module by FullyQualifiedName from absolute path when ModuleVersion=<ModuleVersion>, MaximumVersion=<MaximumVersion>, RequiredVersion=<RequiredVersion>, Guid=<Guid>" -TestCases $guidSuccessCases {
+        param($ModuleVersion, $MaximumVersion, $RequiredVersion, $Guid)
+
+        $modSpec = New-ModuleSpecification -ModuleName $modulePath -ModuleVersion $ModuleVersion -MaximumVersion $MaximumVersion -RequiredVersion $RequiredVersion -Guid $Guid
+
+        $mod = Import-Module -FullyQualifiedName $modSpec -PassThru
+
+        $mod.Name    | Should -Be $moduleName
```

Comment:
This validation code is very similar throughout this script.  It seems you should have a Validate type function instead of the same code repeated